### PR TITLE
Analytical computation of optimal beta per gamma

### PR DESCRIPTION
**Pull request by liranszlak**
### Summary
This PR adds a trainer class that scans gamma values and computes the optimal beta value per gamma, in depth 1 QAOA.

### Details and comments
The class inherits from DepthOneScanTrainer. Tests were added. 
This is based on the computation of beta as per theorem 5 in https://arxiv.org/pdf/2501.16419, with a minor change due to the fact that the paper searches for the optimal beta for minimization of the expected value while in this repository the default is maximizing the problem.

### CODE REVIEW

**Review by eggerdj — CHANGES_REQUESTED** [review]
Thanks a lot Liran, this is looking very good. In addition to the comments on the code here are a few suggestions for the notebook:

- Can you maybe print part of the entries in the result rather then whole result, i.e. the cell with `print(result)`. The output is quite lengthy.
- Would it be possible to have a plot like Fig. 2(a) here: https://arxiv.org/pdf/[2501.16419](https://arxiv.org/pdf/2501.16419)? You could to this by running the `DepthOneScanTrainer` first and then the `DepthOneGammaScanTrainer` second and superimpose both of their plots on the same axis. I think that this would make for a nice and pedagogical figure.

**Comment by eggerdj** [comment]
Oh and one more thing, can you update the version in `training/param_result.py` and the README.md?

- https://github.com/qiskit-community/qaoa_training_pipeline/blob/98c3fcc2c85b80b8729d523c074effee1eafc979/qaoa_training_pipeline/training/param_result.py#L49
- https://github.com/qiskit-community/qaoa_training_pipeline/blob/main/README.md

File: qaoa_training_pipeline/training/parameter_scanner.py [hunk 1]
```diff
@@ -218,3 +221,274 @@ def parse_train_kwargs(self, args_str: Optional[str] = None) -> dict:
...
+
+        Args:
+            evaluator: The evaluator that computes the energy.
+            energy_minimization: Allows us to switch between minimizing the energy or maximizing
+                the energy. The default and assumed convention in this repository is to
+                maximize the energy.
+            qaoa_angles_function: A function to convert optimization parameters into QAOA
+                angles. By default, this is the identity function. Ideally, this argument is
+                an instance of `BaseAnglesFunction` but we allow any callable here that maps
+                optimization parameters to QAOA angles.
```
[the remark this answers is not in the document]
Comment (liranszlak) [reply qaoa_training_pipeline/training/parameter_scanner.py]: Not sure why is should be different from what happens in the DepthOneScanTrainer class? can you explain...?... The gamma and beta parameters are scanned there and passed on to the qaoa_angles_function, and in this class the gamma is scanned and beta is computed analytically, and then they are both passed to the same function (I'm not entirely sure what the function does if it not the identity function, but in terms of compatability, I thought it's exactly the same as in DepthOneScanTrainer). I don't mind removing it of course, just would like to better understand what is the issue, why it works in DepthOneScanTrainer but not in the DepthOneGammaScanTrainer)

File: test/training/test_scanner.py [hunk 2]
```diff
@@ -64,3 +68,87 @@ def test_from_config(self):
         self.assertTrue(isinstance(trainer, DepthOneScanTrainer))
         self.assertTrue(trainer._energy_minimization)
         self.assertTrue(isinstance(trainer.qaoa_angles_function, IdentityFunction))
+
+
+class TestDepthOneGammaScanTrainer(TrainingPipelineTestCase):
```
[the remark this answers is not in the document]
Comment (liranszlak) [reply test/training/test_scanner.py]: done

File: how_tos/optimal_beta_per_gamma_depth_one.ipynb [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+    "cost_op = graph_to_operator(graph, pre_factor=-0.5)\n",
+    "cost_op"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5196c0b3-8dcf-41aa-9906-2dddef38f35d",
+   "metadata": {},
+   "source": [
+    "In the following cell we execute the depth-one parameter scanner. This will compute the energy over a grid of $\\gamma$ and $\\beta$ and pick the maximum energy. The parameters in the returned result are ordered as $[\\beta, \\gamma]$ to match the order of `QAOAAnsatz`."
```
Comment (eggerdj) [inline how_tos/optimal_beta_per_gamma_depth_one.ipynb]: This still needs to be changed, no? How about this?
```suggestion
    "In the following cell we execute the depth-one one-dimensional parameter scanner. This will compute the energy over a grid of $\\gamma$. For each value of $\\gamma$ the optimal $\\beta$ is analytically computed to maximise the energy. The parameters in the returned result are ordered as $[\\beta, \\gamma]$ to match the order of `QAOAAnsatz`."
```

File: test/training/test_scanner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eggerdj) [reply test/training/test_scanner.py]: Thx, looks good.

**Review by eggerdj — CHANGES_REQUESTED** [review]
This looks good and I think we are almost done. A few more edits to the notebook and crucially I think we need to raise on some of the edge cases that this method does not yet support. If there is a need for them (like the angles function or non-standard mixers) we can do this in future PRs.

**Review by eggerdj — COMMENTED** [review]
Looks good to me. Can you resolve the conflicts and then we merge?

**Review by eggerdj — APPROVED** [review]
LGTM!
